src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpClientJsonExtensions.cs

```diff
@@ -0,0 +1,129 @@
+            FromJsonAsyncCore(getMethod, client, requestUri, static (stream, options, cancellation) => JsonSerializer.DeserializeAsync(stream, options, cancellation), jsonTypeInfo, cancellationToken);
+
+        private static Task<TValue?> FromJsonAsyncCore<TValue, TJsonOptions>(
+            Func<HttpClient, Uri?, CancellationToken, Task<HttpResponseMessage>> getMethod,
+            HttpClient client,
+            Uri? requestUri,
+            Func<Stream, TJsonOptions, CancellationToken, ValueTask<TValue?>> deserializeMethod,
+            TJsonOptions jsonOptions,
+            CancellationToken cancellationToken)
+        {
+            if (client is null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+
+            TimeSpan timeout = client.Timeout;
+
+            // Create the CTS before the initial SendAsync so that the SendAsync counts against the timeout.
+            CancellationTokenSource? linkedCTS = null;
+            if (timeout != Timeout.InfiniteTimeSpan)
+            {
+                linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                linkedCTS.CancelAfter(timeout);
+            }
+
+            // We call SendAsync outside of the async Core method to propagate exception even without awaiting the returned task.
+            Task<HttpResponseMessage> responseTask;
+            try
+            {
+                // Intentionally using cancellationToken instead of the linked one here as HttpClient will enforce the Timeout on its own for this part
+                responseTask = getMethod(client, requestUri, cancellationToken);
+            }
+            catch
+            {
+                linkedCTS?.Dispose();
+                throw;
+            }
+
+            bool usingResponseHeadersRead = !ReferenceEquals(getMethod, s_deleteAsync);
+
+            return Core(client, responseTask, usingResponseHeadersRead, linkedCTS, deserializeMethod, jsonOptions, cancellationToken);
+
+            static async Task<TValue?> Core(
+                HttpClient client,
+                Task<HttpResponseMessage> responseTask,
+                bool usingResponseHeadersRead,
+                CancellationTokenSource? linkedCTS,
+                Func<Stream, TJsonOptions, CancellationToken, ValueTask<TValue?>> deserializeMethod,
+                TJsonOptions jsonOptions,
+                CancellationToken cancellationToken)
+            {
+                try
+                {
+                    using HttpResponseMessage response = await responseTask.ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+
+                    Debug.Assert(client.MaxResponseContentBufferSize is > 0 and <= int.MaxValue);
+                    int contentLengthLimit = (int)client.MaxResponseContentBufferSize;
+
+                    HttpContent content = response.Content ?? throw new ArgumentNullException(nameof(content));
```

Comment:
In what situation will response.Content be null?
https://github.com/dotnet/runtime/blob/58614a0f6da787c9d6d914b99564f066d13f69ed/src/libraries/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L42-L44

Reply:
In practice probably never, but as far as I understand this project ships as a NuGet package that targets Framework as well, where `Content` may still return null.

Reply:
If we really care about that situation, where some HttpClientHandler is effectively erroneously returning null from Content, and we really want a different exception to be thrown than NullReferenceException, it shouldn't be ArgumentNullException.